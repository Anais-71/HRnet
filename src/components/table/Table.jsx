import React, { useState, useMemo } from 'react'
import './table.css'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {
  faSearch,
  faSortUp,
  faSortDown,
  faBackwardStep,
  faForwardStep,
  faChevronRight,
  faChevronLeft,
} from '@fortawesome/free-solid-svg-icons'

const Table = ({ columns, data }) => {
  const [searchTerm, setSearchTerm] = useState('')
  const [sortConfig, setSortConfig] = useState({ key: null, direction: 'asc' })
  const [entriesPerPage, setEntriesPerPage] = useState(10)
  const [currentPage, setCurrentPage] = useState(1)

  const sortedData = useMemo(() => {
    let sorted = [...data]
    if (sortConfig.key) {
      sorted.sort((a, b) => {
        const aValue = a[sortConfig.key]
        const bValue = b[sortConfig.key]

        if (typeof aValue === 'string' && typeof bValue === 'string') {
          return sortConfig.direction === 'asc'
            ? aValue.localeCompare(bValue)
            : bValue.localeCompare(aValue)
        } else if (typeof aValue === 'number' && typeof bValue === 'number') {
          return sortConfig.direction === 'asc'
            ? aValue - bValue
            : bValue - aValue
        }
        return 0
      })
    }
    return sorted
  }, [data, sortConfig])

  const filteredData = useMemo(() => {
    return sortedData.filter((item) =>
      Object.values(item).some((value) =>
        value.toString().toLowerCase().includes(searchTerm.toLowerCase()),
      ),
    )
  }, [sortedData, searchTerm])

  const displayedData = filteredData.slice(
    (currentPage - 1) * entriesPerPage,
    currentPage * entriesPerPage,
  )

  const totalEntries = filteredData.length
  const totalPages = Math.ceil(totalEntries / entriesPerPage)

  const handleSort = (key) => {
    setSortConfig((prev) => ({
      key,
      direction: prev.direction === 'asc' && prev.key === key ? 'desc' : 'asc',
    }))
  }

  const handleSearch = (e) => {
    setSearchTerm(e.target.value)
    setCurrentPage(1)
  }

  const handleEntriesPerPage = (e) => {
    setEntriesPerPage(Number(e.target.value))
    setCurrentPage(1)
  }

  const changePage = (page) => {
    if (page >= 1 && page <= totalPages) {
      setCurrentPage(page)
    }
  }

  const startEntry = (currentPage - 1) * entriesPerPage + 1
  const endEntry = Math.min(currentPage * entriesPerPage, totalEntries)

  return (
    <div className="container">
      <div className="header__controls">
        <div
          className="header__controls--search"
          data-testid="search-input-container"
        >
          <FontAwesomeIcon
            icon={faSearch}
            className="header__controls--search--icon"
          />
          <input
            type="text"
            placeholder="Search..."
            value={searchTerm}
            onChange={handleSearch}
            className="header__controls--search--input"
            data-testid="search-input"
          />
        </div>
        <div className="header__controls--pages">
          Show
          <select
            value={entriesPerPage}
            onChange={handleEntriesPerPage}
            className="header__controls--pages--select"
            data-testid="entries-per-page-select"
          >
            {[10, 25, 50, 100].map((size) => (
              <option key={size} value={size}>
                {size}
              </option>
            ))}
          </select>
          entries
        </div>
      </div>

      <table className="table">
        <thead className="table__header">
          <tr>
            {columns.map((column) => (
              <th
                className="table__header--row"
                key={column.accessor}
                onClick={() => handleSort(column.accessor)}
                data-testid={`sort-column-${column.accessor}`}
              >
                {column.Header}
                {sortConfig.key === column.accessor && (
                  <span className="sort-icon">
                    <FontAwesomeIcon
                      icon={
                        sortConfig.direction === 'asc' ? faSortUp : faSortDown
                      }
                    />
                  </span>
                )}
              </th>
            ))}
          </tr>
        </thead>
        <tbody className="table__content">
          {displayedData.map((item, index) => (
            <tr className="table__row" key={index} data-testid="table-row">
              {columns.map((column) => (
                <td className="table__col" key={column.accessor}>
                  {item[column.accessor]}
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>

      <div className="table__footer">
        <p className="table__footer--entries">
          Showing {startEntry} to {endEntry} of {totalEntries} entries
        </p>

        <div
          className="table__footer--pagination"
          data-testid="pagination-buttons"
        >
          <button
            className="table__footer--pagination--btn"
            onClick={() => changePage(1)}
            disabled={currentPage === 1}
            data-testid="first-page-btn"
          >
            <FontAwesomeIcon icon={faBackwardStep} />
          </button>
          <button
            className="table__footer--pagination--btn"
            onClick={() => changePage(currentPage - 1)}
            disabled={currentPage === 1}
            data-testid="prev-page-btn"
          >
            <FontAwesomeIcon icon={faChevronLeft} />
          </button>

          {[...Array(totalPages)].map((_, index) => (
            <button
              key={index + 1}
              onClick={() => changePage(index + 1)}
              className={`table__footer--pagination--btn ${
                currentPage === index + 1 ? 'active' : ''
              }`}
              data-testid={`page-btn-${index + 1}`}
            >
              {index + 1}
            </button>
          ))}

          <button
            className="table__footer--pagination--btn"
            onClick={() => changePage(currentPage + 1)}
            disabled={currentPage === totalPages}
            data-testid="next-page-btn"
          >
            <FontAwesomeIcon icon={faChevronRight} />
          </button>
          <button
            className="table__footer--pagination--btn"
            onClick={() => changePage(totalPages)}
            disabled={currentPage === totalPages}
            data-testid="last-page-btn"
          >
            <FontAwesomeIcon icon={faForwardStep} />
          </button>
        </div>
      </div>
    </div>
  )
}

export default Table
