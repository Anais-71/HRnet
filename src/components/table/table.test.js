import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import React from 'react'
import Table from './Table'

/**
 * Test the search functionality in the Table component.
 * When the user types in the search input, the table filters data accordingly.
 */
test('When I use the search input, then search filters data based on input', async () => {
  const columns = [{ Header: 'Name', accessor: 'name' }]
  const data = [{ name: 'Alice' }, { name: 'Bob' }, { name: 'Charlie' }]

  render(<Table columns={columns} data={data} />)

  const searchInput = screen.getByTestId('search-input')

  // Simulate typing a search term into the input field
  fireEvent.change(searchInput, { target: { value: 'Alice' } })

  // Verify that only 'Alice' is visible
  expect(screen.getByText('Alice')).toBeInTheDocument()
  expect(screen.queryByText('Bob')).not.toBeInTheDocument()
})

/**
 * Test sorting functionality in the Table component.
 * When the user clicks on a column header, the table sorts data by that column.
 */
test('When I click the sort button, then the table sorts the data', async () => {
  const columns = [{ Header: 'Name', accessor: 'name' }]
  const data = [{ name: 'Alice' }, { name: 'Bob' }, { name: 'Charlie' }]

  render(<Table columns={columns} data={data} />)

  const sortColumn = screen.getByTestId('sort-column-name')

  // Simulate clicking the sort column
  fireEvent.click(sortColumn)

  // Verify that the first row is 'Alice' after sorting
  const firstRow = screen.getByText('Alice')
  expect(firstRow).toBeInTheDocument()

  // Simulate clicking again to reverse the sort order
  fireEvent.click(sortColumn)

  // Verify that the first row is 'Charlie' after sorting in reverse
  const firstRowAfterSort = screen.getByText('Charlie')
  expect(firstRowAfterSort).toBeInTheDocument()
})

/**
 * Test entries per page functionality in the Table component.
 * When the user changes the entries per page option, the table updates to display the correct number of entries.
 */
test('When I change entries per page, then the number of entries displayed changes accordingly', async () => {
  const columns = [{ Header: 'Name', accessor: 'name' }]
  const data = Array.from({ length: 50 }, (_, i) => ({ name: `User ${i + 1}` }))

  render(<Table columns={columns} data={data} />)

  const entriesPerPageSelect = screen.getByTestId('entries-per-page-select')

  // Simulate selecting 25 entries per page
  fireEvent.change(entriesPerPageSelect, { target: { value: '25' } })

  // Verify that 25 rows are displayed
  const rows = screen.getAllByTestId('table-row')
  expect(rows.length).toBe(25)
})

/**
 * Test pagination functionality in the Table component.
 * When the user clicks the pagination buttons, the table updates to show the correct page.
 */
test('When I click on pagination buttons, then the page changes and entries update accordingly', async () => {
  const columns = [{ Header: 'Name', accessor: 'name' }]
  const data = Array.from({ length: 25 }, (_, i) => ({ name: `User ${i + 1}` }))

  render(<Table columns={columns} data={data} />)

  // Verify that the first page is active by default
  const pageOneButton = screen.getByTestId('page-btn-1')
  expect(pageOneButton).toHaveClass('active')

  const nextPageButton = screen.getByTestId('next-page-btn')
  fireEvent.click(nextPageButton)

  // Verify that the second page is active after clicking 'Next'
  const pageTwoButton = screen.getByTestId('page-btn-2')
  expect(pageTwoButton).toHaveClass('active')

  // Verify that the first page is no longer active
  expect(pageOneButton).not.toHaveClass('active')

  // Verify that data for page 2 (e.g., User 11) is displayed
  expect(screen.getByText('User 11')).toBeInTheDocument()
})

/**
 * Test first and last page navigation in the Table component.
 * When the user clicks the first or last page buttons, the table should update correctly.
 */
test('When I go to the first and last page, then the pagination buttons behave correctly', async () => {
  const columns = [{ Header: 'Name', accessor: 'name' }]
  const data = Array.from({ length: 50 }, (_, i) => ({ name: `User ${i + 1}` }))

  render(<Table columns={columns} data={data} />)

  const lastPageButton = screen.getByTestId('last-page-btn')
  fireEvent.click(lastPageButton)

  // Verify that the last page is active
  const lastPageButtonActive = screen.getByTestId('page-btn-5') // assuming there are 5 pages
  expect(lastPageButtonActive).toHaveClass('active')

  // Go to the first page
  const firstPageButton = screen.getByTestId('first-page-btn')
  fireEvent.click(firstPageButton)

  // Verify that the first page is active
  const firstPageButtonActive = screen.getByTestId('page-btn-1')
  expect(firstPageButtonActive).toHaveClass('active')
})

/**
 * Test the behavior when no data matches the search term.
 * If no data matches, no rows should be displayed in the table.
 */
test('When no data matches the search term, then no rows are displayed', async () => {
  const columns = [{ Header: 'Name', accessor: 'name' }]
  const data = [{ name: 'Alice' }, { name: 'Bob' }, { name: 'Charlie' }]

  render(<Table columns={columns} data={data} />)

  const searchInput = screen.getByTestId('search-input')

  // Search for a non-existent term
  fireEvent.change(searchInput, { target: { value: 'Zoe' } })

  // Verify that no rows are displayed
  const rows = screen.queryAllByTestId('table-row')
  expect(rows.length).toBe(0)
})

/**
 * Snapshot test for the Table component.
 * The rendered table is compared to a saved snapshot to ensure consistent output.
 */
test('Table matches the snapshot', () => {
  const columns = [{ Header: 'Name', accessor: 'name' }]
  const data = [{ name: 'Alice' }, { name: 'Bob' }, { name: 'Charlie' }]

  const { asFragment } = render(<Table columns={columns} data={data} />)

  // Compare the rendered output to the saved snapshot
  expect(asFragment()).toMatchSnapshot()
})
