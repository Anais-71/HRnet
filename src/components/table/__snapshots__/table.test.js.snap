// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="header__controls"
    >
      <div
        class="header__controls--search"
        data-testid="search-input-container"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-magnifying-glass header__controls--search--icon"
          data-icon="magnifying-glass"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
            fill="currentColor"
          />
        </svg>
        <input
          class="header__controls--search--input"
          data-testid="search-input"
          placeholder="Search..."
          type="text"
          value=""
        />
      </div>
      <div
        class="header__controls--pages"
      >
        Show
        <select
          class="header__controls--pages--select"
          data-testid="entries-per-page-select"
        >
          <option
            value="10"
          >
            10
          </option>
          <option
            value="25"
          >
            25
          </option>
          <option
            value="50"
          >
            50
          </option>
          <option
            value="100"
          >
            100
          </option>
        </select>
        entries
      </div>
    </div>
    <table
      class="table"
    >
      <thead
        class="table__header"
      >
        <tr>
          <th
            class="table__header--row"
            data-testid="sort-column-name"
          >
            Name
          </th>
        </tr>
      </thead>
      <tbody
        class="table__content"
      >
        <tr
          class="table__row"
          data-testid="table-row"
        >
          <td
            class="table__col"
          >
            Alice
          </td>
        </tr>
        <tr
          class="table__row"
          data-testid="table-row"
        >
          <td
            class="table__col"
          >
            Bob
          </td>
        </tr>
        <tr
          class="table__row"
          data-testid="table-row"
        >
          <td
            class="table__col"
          >
            Charlie
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="table__footer"
    >
      <p
        class="table__footer--entries"
      >
        Showing 1 to 3 of 3 entries
      </p>
      <div
        class="table__footer--pagination"
        data-testid="pagination-buttons"
      >
        <button
          class="table__footer--pagination--btn"
          data-testid="first-page-btn"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-backward-step "
            data-icon="backward-step"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M267.5 440.6c9.5 7.9 22.8 9.7 34.1 4.4s18.4-16.6 18.4-29l0-320c0-12.4-7.2-23.7-18.4-29s-24.5-3.6-34.1 4.4l-192 160L64 241 64 96c0-17.7-14.3-32-32-32S0 78.3 0 96L0 416c0 17.7 14.3 32 32 32s32-14.3 32-32l0-145 11.5 9.6 192 160z"
              fill="currentColor"
            />
          </svg>
        </button>
        <button
          class="table__footer--pagination--btn"
          data-testid="prev-page-btn"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-left "
            data-icon="chevron-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l192 192c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L77.3 256 246.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-192 192z"
              fill="currentColor"
            />
          </svg>
        </button>
        <button
          class="table__footer--pagination--btn active"
          data-testid="page-btn-1"
        >
          1
        </button>
        <button
          class="table__footer--pagination--btn"
          data-testid="next-page-btn"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-right "
            data-icon="chevron-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M310.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L242.7 256 73.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"
              fill="currentColor"
            />
          </svg>
        </button>
        <button
          class="table__footer--pagination--btn"
          data-testid="last-page-btn"
          disabled=""
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-forward-step "
            data-icon="forward-step"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M52.5 440.6c-9.5 7.9-22.8 9.7-34.1 4.4S0 428.4 0 416L0 96C0 83.6 7.2 72.3 18.4 67s24.5-3.6 34.1 4.4l192 160L256 241l0-145c0-17.7 14.3-32 32-32s32 14.3 32 32l0 320c0 17.7-14.3 32-32 32s-32-14.3-32-32l0-145-11.5 9.6-192 160z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
